### PR TITLE
[Dashboard] Optimize and backpressure actor_head.py

### PR DESCRIPTION
**Pull request by rkooo567**
Signed-off-by: SangBin Cho <rkooo567@gmail.com>

## Why are these changes needed?

This optimizes the actor head CPU usage and guarantees a stable API response from the dashboard under lots of actor events published to drivers. The below script is used for testing, and I could reproduce the same level of delay as many_nodes_actor_test (250 nodes + 10k actors)

```python3
import ray
ray.init()

@ray.remote
class A:
    pass
import time
while True:
    a = [A.remote() for _ in range(100)]
    time.sleep(0.1)
    del a
```

- Remove unnecessary fields that are not used (including that causes high CPU usage)
- Add backpressure to event processing time (1000/s) so that the dashboard will have enough CPU to respond API requests
- Batch processing actor events and minimize the context switching time
- Stop using immutable dict that causes high copy overhead

The profiling result after this PR is here; https://github.com/ray-project/ray/pull/29580#issuecomment-1306310604

After this PR, most of overhead is just pure MessageToDict. We can drastically optimize this by delaying the processing of protobuf on query time. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/datacenter.py [hunk 1]
```diff
@@ -39,11 +42,7 @@ class DataSource:
     # {node id hex(str): worker list}
     node_workers = Dict()
     # {node id hex(str): {actor id hex(str): actor table data}}
-    node_actors = Dict()
-    # {job id hex(str): worker list}
-    job_workers = Dict()
```
Comment (rickyyx) [inline dashboard/datacenter.py:44]: Why are these removed?

File: dashboard/datacenter.py [hunk 2]
```diff
@@ -25,7 +28,7 @@ class DataSource:
     node_physical_stats = Dict()
     # {actor id hex(str): actor table data(dict of ActorTableData
     # in gcs.proto)}
-    actors = Dict()
+    actors = MutableNotificationDict()
```
Comment (rickyyx) [inline dashboard/datacenter.py:31]: what's diff from these two?

**Review by rickyyx — COMMENTED** [review]
~~Could you add some more details to the description for the changes and issues? I might be missing the context here, thank you!~~
Saw the slack thread.

**Comment by ericl** [comment]
Any before/after profiling comparisons?

**Comment by rkooo567** [comment]
@ericl oh forgot to update it. Let me share it here. Note that for the "after" flamegraph, it includes overhead of /actors endpoint (so I just didn't include that page to the screenshot)

Before
Screen Shot 2022-11-07 at 6 27 16 AM

After
Screen Shot 2022-11-07 at 6 25 54 AM

File: dashboard/modules/actor/actor_head.py [hunk 3]
```diff
@@ -153,39 +138,53 @@ def process_actor_data_from_pubsub(actor_id, actor_table_data):
...
+                start = time.monotonic()
+                for actor_id, actor_table_data in published:
+                    if actor_id is not None:
+                        # Convert to lower case hex ID.
+                        actor_id = actor_id.hex()
+                        process_actor_data_from_pubsub(actor_id, actor_table_data)
+                    total_published_events += 1
+                    # Backpressure the # of actor event process per second.
+                    if total_published_events % ACTORS_TO_PROCESS_PER_SECOND == 0:
+                        await asyncio.sleep(1)
```
Comment (ericl) [inline dashboard/modules/actor/actor_head.py:176]: Could we just yield the loop without sleeping, or sleep a smaller amount?

**Review by architkulkarni — APPROVED** [review]
`test_snapshot` changes look good to me, stamping as codeowner

File: dashboard/modules/actor/actor_head.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/modules/actor/actor_head.py:176]: We can also adjust `ACTORS_TO_PROCESS_PER_SECOND`? What's the desired number you think? (rn, it can do up to 1000 events/s)

**Review by rkooo567 — COMMENTED** [review]
@rickyyx I will update the description. I also forgot to add comments which explains the PR...

File: dashboard/modules/actor/actor_head.py (hunk 3 above)
Comment (ericl) [reply dashboard/modules/actor/actor_head.py:176]: We don't need to throttle this right? The only problem is if it starves other events from being processed.

Actually, I'd favor just removing this limit, since the number of actor events is finite and we yield at some point right? A couple seconds extra of dashboard latency right when starting a lot of actors is probably no big deal, or am I missing something?

File: dashboard/utils.py [hunk 4]
```diff
@@ -439,6 +439,52 @@ def __repr__(self):
...
+class MutableNotificationDict(dict, MutableMapping):
+    """A simple descriptor for dict type to notify data changes.
+    :note: Only the first level data report change.
+    """
+
+    ChangeItem = namedtuple("DictChangeItem", ["key", "value"])
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.signal = Signal(self)
```
Comment (ericl) [inline dashboard/utils.py:451]:
```suggestion
        self._signal = Signal(self)
```

File: dashboard/modules/actor/actor_head.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/modules/actor/actor_head.py:176]: Let's talk about this in 1:1 tomorrow offline? I prefer to throttle it unless you are really strongly against it because this makes a really huge difference in user experience (I can demonstrate this later through videos). I have 3 reasons why I prefer to throttle it.

1. In small-scale clusters, this has no impact. 1-second delay after processing 1000 events is like nothing in small scale cluster.
2. This just doesn't starve other events. It starves all the API responses, and all the API (including node + job) takes about 5~10 seconds to reply. With throttling, they pretty much respond within 1 second at the same scale.
3. Since the dashboard is a user-facing application, we should have a higher standard for API response time. 5~10 seconds response time for all APIs is not acceptable. Everyone's going to just think the dashboard. 

I think the only con is the actor information can be stale (at very large scale), but it is much better than the whole dashboard being unresponsive. 

I can also disable it after I optimize this processing path as I talked about in the slack channel, and if the response is time is much more reasonable.

**Review by rickyyx — APPROVED** [review]

File: dashboard/modules/actor/actor_head.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/modules/actor/actor_head.py]: Addressed based on our discussion (sleep as much as processing time). I found after applying this policy, the load on the dashboard is really good

**Comment by rkooo567** [comment]
Made additional changes.

1. Increase the max actors to cache 1k -> 10K. It makes no sense it is 1K given our scalability limit is 10K concurrent actors. I even feel like it should something like 20K.
2. Reduce the `cleanup_actors` frequency so that each call will be shorter (so that it won't occupy the main thread long time).
3. Reduce the batch size for the same reason ^

**Comment by rickyyx** [comment]
> * Increase the max actors to cache 1k -> 10K. It makes no sense it is 1K given our scalability limit is 10K concurrent actors. I even feel like it should something like 20K.

What will be a measurement to see if this limit doesn't have a siginificnant impact and could be adjusted?

**Comment by rkooo567** [comment]
> What will be a measurement to see if this limit doesn't have a siginificnant impact and could be adjusted?

I think the workload I am running (stressful actor workload) is the one that can do it. But I wonder if this can impact the prod (e.g., snapshot API becomes slow or broken due to the large API response)
